streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2002,58 +2002,40 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
-          fiber.interrupt.as(Take.End)
-        case _ =>
-          fiber.join.map(_.map(_.map(mapper))).map(_.map((_, (left, right, None))))
-      }
+  )(l: O => O3, r: O2 => O3): ZStream[R1, E1, O3] =
+    ZStream {
+      import TerminationStrategy.{ Left => L, Right => R, Either => E }
 
-    combineChunks(that)((false, false, Option.empty[Loser])) {
-      case ((leftDone, rightDone, loser), left, right) =>
-        if (leftDone) {
-          right.map(c => (c.map(r), (leftDone, rightDone, None))).run
-        } else if (rightDone) {
-          left.map(c => (c.map(l), (leftDone, rightDone, None))).run
-        } else {
-          val result = loser match {
-            case None               => race(left.run, right.run)
-            case Some(Left(loser))  => race(loser.join, right.run)
-            case Some(Right(loser)) => race(left.run, loser.join)
-          }
-          result.flatMap {
-            case (exit, loser) =>
-              exit.foldM(
-                Cause.sequenceCauseOption(_) match {
-                  case Some(e) =>
-                    loser.merge.interrupt.as(Exit.halt(e.map(Some(_))))
-                  case None =>
-                    loser.fold(
-                      termination[E, O](leftDone, true, l),
-                      termination[E1, O2](true, rightDone, r)
-                    )
-                },
-                chunk => ZIO.succeedNow(Exit.succeed((chunk, (leftDone, rightDone, Some(loser)))))
-              )
-          }
+      for {
+        queue   <- Queue.unbounded[Take[E1, O3]].toManaged(_.shutdown)
+        done    <- RefM.make[Option[Boolean]](None).toManaged_
+        chunksL <- self.process
+        chunksR <- that.process
+        handler = (pull: Pull[R1, E1, O3], terminate: Boolean) => {
+          pull
+            .foldCause(c => Left(Cause.sequenceCauseOption(c)), Right(_))
+            .flatMap { either =>
+              done.modify {
+                (_, either) match {
+                  case (state @ Some(true), _) =>
+                    ZIO.succeedNow((false, state))
+                  case (state, Right(chunk)) =>
+                    queue.offer(Exit.succeed(chunk)).as((true, state))
+                  case (_, Left(Some(cause))) =>
+                    queue.offer(Exit.halt(cause.map(Some(_)))).as((false, Some(true)))
+                  case (option, Left(None)) if terminate || option.isDefined =>
+                    queue.offer(Exit.fail(None)).as((false, Some(true)))
+                  case (None, Left(None)) =>
+                    ZIO.succeedNow((false, Some(false)))
+                }
+              }
+            }
+            .repeat(Schedule.doWhileEquals(true))
         }
+        fiberL <- handler(chunksL.map(_.map(l)), List(L, E).contains(strategy)).forkManaged
+        fiberR <- handler(chunksR.map(_.map(r)), List(R, E).contains(strategy)).forkManaged
+      } yield queue.take.flatMap(IO.done(_)).tapError(_ => fiberL.interrupt *> fiberR.interrupt)
```

Comment:
I don't think fibers should be interrupted on error since a stream can resume after producing an error. But if the stream is interrupted from the outside that's another story. However since upstream can terminate, why not query the `done` flag then:
```suggestion
      } yield done.get.flatMap(done =>
        if (done.contains(true)) Take.end 
        else queue.take.flatMap(IO.done(_)).onInterrupt(_ => fiberL.interrupt *> fiberR.interrupt))
```

Reply:
The `onInterrupt` is actually not needed either, because the fibers are forked with `forkManaged`, so that will take care of the interruption when the ZManaged scope ends.

Reply:
@regiskuckaertz wouldn't that change throw away elements in the queue that happened before the end of the stream and haven't been pulled yet?

Reply:
Well, replaced the queue with a Handoff anyway

Reply:
@luis3m right you are, I guess you could `poll` the queue in the consequent instead of just `Take.end` 🤔

Reply:
@regiskuckaertz thanks to your other suggestion it isn't a queue anymore 😅